### DISCUSSION TITLE
F# in dotnet 8?

### DISCUSSION THREAD

**Discussion by nojaf** [dotnet/fsharp#14619] [Q&A] [answered]
Hello,

I would like to know what code is present in the [`dotnet 8` nightlies](https://github.com/dotnet/installer#table).

When something is merged to main, I always very much struggle to know when this will finally be available in an SDK.
What goes to the next `7.0.200`, `7.0.300`, `8.0.100`?

I downloaded the `8.0.100-alpha.1.23063.11` today and I can't really say what the last commit was.
It looks recent, but I just can't tell for sure. Maybe I'm missing something here.

Anyway, I think others might have this problem as well. I know @edgarfgp got excited when the `7.0.102` was released, hoping that https://github.com/dotnet/fsharp/pull/13839 would be available, but it was not the case.

Lastly, I would very much like to have an SDK with the code from `main` in it, if that does not already exist. Some more regression testing could be done. Things like https://github.com/dotnet/fsharp/issues/14598 would surface a lot quicker, instead of them being detected by accident in a single code-base.

Besides regressions, I would like to set something up something that tests out the results of certain flags like https://github.com/dotnet/fsharp/pull/14494. Having an SDK with the latest main could make that a lot easier. The thing I have in mind is a GitHub Action, that downloads the latest nightly, clones a couple of repositories and builds them. Running that on a weekly basis would boost confidence in new features and the overall next iteration of the compiler.

**Comment by kerams** [comment, marked as answer]
- Download https://aka.ms/dotnet/8.0.1xx/daily/productCommit-win-x64.txt
- Copy the commit hash for `sdk_commit`
- Open `https://github.com/dotnet/sdk/blob/{SDK_COMMIT_FROM_ABOVE}/eng/Version.Details.xml`
- Copy the commit hash for `<Dependency Name="Microsoft.FSharp.Compiler"`
- Open `https://github.com/dotnet/fsharp/commit/{FSHARP_COMMIT_FROM_ABOVE}`

That gets you the last F# commit included in the .NET 8 dailies. You can see that it's ~2 weeks behind the HEAD in main, since changes take some time to bubble up through multiple repos.

> When something is merged to main, I always very much struggle to know when this will finally be available in an SDK.
What goes to the next 7.0.200, 7.0.300, 8.0.100

I reckon that unless a bug fix is manually cherry picked and merged into a release branch, commits go towards the next major version only.

As for VS tooling, changes tend to land in a preview around 1 or 2 months after having been merged.

**Comment by vzarytovskii** [comment]
@baronfel may have more info about how .NET 8 insertions work

**Comment by nojaf** [comment]
Thanks @kerams! Does this mean something needs to go to `dev17.5` before it goes into `dotnet 8`?
If that is the case, can't this be hooked up to `main` instead?

**Reply by kerams** [reply]
It should be `dev17.6` now. dotnet-bot opens [PRs with new commits from main automatically](https://github.com/dotnet/fsharp/pull/14611). I'm guessing this setup gives the team more fine-grained control over insertions.

**Reply by baronfel** [reply]
I think the main issue I see is that the current flow to SDK main (where 8.0 nightlies come from) is from the 17.5 branch here. This means that anything that doesn't meet the 17.5 ship bar is blocked from any reasonable contributor use for quite some time.

I'd love to see either a 17.8 (guessing the name of the November VS release) branch here with flow to SDK main, or a flow from main here to SDK main until the correct 17.8 main is opened here in the fullness of time.

**Reply by vzarytovskii** [reply]
@baronfel where does 17.6 channel flow in? `7.0.300`?

**Reply by kerams** [reply]
Interesting, I expected the source of the flow to main would always match the latest F# release branch.

**Reply by baronfel** [reply]
Yes, which I believe has just been created on the SDK.

SDK main is almost always 'next major version'. The only time this isn't the case is around September when we make the 'release/next major version' branch for the RCs and eventually the stables to branch off from.

**Reply by baronfel** [reply]
@kerams the blocker here is the release/cadence difference between the fsharp and SDK teams I think.

I think fsharp is very focused on 'next VS release', which is why main doesn't go anywhere in particular and specific release branches are used to cherry pick specific changes from main for those releases.

SDK has both a 'next SDK feature band' (same cadence as VS Release) focus but it crucially has a 'next major SDK version' focus, which I think isn't so much a thing in this repo.

**Reply by baronfel** [reply]
Note that I'm not trying to make a value judgement or anything, just attempting to describe the high level difference in approach

**Comment by edgarfgp** [comment]
From contributor point of view. Having to wait for several months (in some cases) is a little discorauging and also the way to setup the nightly net8 builds is long tedious and confusing. I appreciate all of this has a lot of complexity so I not compalining  just giving my observations :)

**Comment by nojaf** [comment]
Hi @vzarytovskii, the latest commit to the nightlies it https://github.com/dotnet/fsharp/commit/e391694876e891d2cdc4089e62fe7b1138d8294e for some time now.
Can `release/dev17.6` be hooked to the nightlies, please?

**Reply by vzarytovskii** [reply]
From what I can see, insertions are still going to SDK - https://github.com/dotnet/sdk/pulls?q=is%3Apr+sort%3Aupdated-desc+dotnet%2Ffsharp

I don't know how nuget publishing is working for FCS  though.

Cc @baronfel @KevinRansom.

**Reply by nojaf** [reply]
I would guess that those insertions are coming from `release/dev17.5` right now.

> I don't know how nuget publishing is working for FCS though.

On that topic, I would very much like to have another public FCS preview.
Right now the community has no way of anticipating the changes for the `7.0.200` release.
If something isn't working (as was the case with `7.0.100`), the is no time to respond to this in time.

**Reply by vzarytovskii** [reply]
Also, maybe @mmitche knows how nugets are published?

**Reply by vzarytovskii** [reply]
I can guess that they're published from the 17.5 until it's released? And then it'll be 17.6? Not sure why it's not both/all darc channels.

**Reply by vzarytovskii** [reply]
I see the channels set up in the following way:
```
(4541) https://github.com/dotnet/fsharp @ release/dev17.5 -> VS 17.5
(4621) https://github.com/dotnet/fsharp @ release/dev17.6 -> VS 17.6
```

**Reply by mmitche** [reply]
The insertions into the SDK are coming from the 17.6 channel, which is populated by fsharp builds from the 17.6 branch. Both these and 17.5 builds end up in the dotnet-tools feed https://github.com/dotnet/sdk/blob/main/NuGet.config#L27.

I see the latest build was a little after midnight PST.

**Reply by vzarytovskii** [reply]
> The insertions into the SDK are coming from the 17.6 channel, which is populated by fsharp builds from the 17.6 branch. Both these and 17.5 builds end up in the dotnet-tools feed https://github.com/dotnet/sdk/blob/main/NuGet.config#L27.
> 
> I see the latest build was a little after midnight PST. 

Thanks, Matt! @nojaf can you please check said feed and see if the packages are there (I'm on vacation this and next week, and won't have access to my laptop until Monday).

**Reply by nojaf** [reply]
Ok, there are two things here:
1. When following the steps @kerams provided, I end up in https://github.com/dotnet/fsharp/commit/e391694876e891d2cdc4089e62fe7b1138d8294e for some time now. This leads me to believe `17.5` is still wired to the `dotnet 8` nightlies and not `17.6`.
2. I was more talking about the public NuGet feed in terms of the FCS package.

**Reply by vzarytovskii** [reply]
I don't know what ends up in 8, probability @baronfel can help here?

Regarding public versions - for now they're tied to SDK releases, we'll need to think how shall we publish previews more often.

**Reply by baronfel** [reply]
Right now, anything that flows to dotnet/sdk `main` would be part of .NET 8. Currently the 17.6 branch from this repo is flowing to main. 

I can confirm that nightlies of FCS are being pushed to the dotnet-tools feed, the current versioning scheme is 43.7.300-preview.xxxx.  That matches with the idea that 17.6 versions are flowing there.

One thing I'm not sure of is how/if packages from the `main` branch of dotnet/fsharp end up in the `dotnet-tools` feed at all (or otherwise in any end-user-consumable form). Right now there's no channel/branch for 'things that will never end up in .NET 7, only in .NET 8' for this repo, as far as I can tell.

**Reply by vzarytovskii** [reply]
> Right now, anything that flows to dotnet/sdk `main` would be part of .NET 8. Currently the 17.6 branch from this repo is flowing to main. 
> 
> I can confirm that nightlies of FCS are being pushed to the dotnet-tools feed, the current versioning scheme is 43.7.300-preview.xxxx.  That matches with the idea that 17.6 versions are flowing there.
> 
> One thing I'm not sure of is how/if packages from the `main` branch of dotnet/fsharp end up in the `dotnet-tools` feed at all (or otherwise in any end-user-consumable form). Right now there's no channel/branch for 'things that will never end up in .NET 7, only in .NET 8' for this repo, as far as I can tell.

Thanks. F# `main` is not flowing anywhere right now, or rather not inserted, so I don't think packages are even published anywhere. Also, versioning scheme in main is usually behind latest release branch.

**Comment by Xyncgas** [comment]
What's coming in .net 8
